### ISSUE TITLE
[QUERY] Azure.Identity.CredentialDiagnosticScope.FailWrapAndThrow exception after migrating to Azure.Identity

### ISSUE THREAD

**Issue by shahriak** [Azure/azure-sdk-for-net#23357]
**Query/Question**
How can we help?
We are in the process of doing ADAL to MSAL migration for Managed Identity (MSI) and followed both of these docs in order to accomplish this:
https://docs.microsoft.com/en-us/dotnet/api/overview/azure/app-auth-migration
https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/keyvault/Azure.Security.KeyVault.Certificates/MigrationGuide.md

We are seeing the following error message after deploying the changes:

```
 at Azure.Identity.CredentialDiagnosticScope.FailWrapAndThrow(Exception ex)
   at Azure.Identity.ManagedIdentityCredential.GetTokenImplAsync(Boolean async, TokenRequestContext requestContext, CancellationToken cancellationToken)
   at Azure.Identity.ManagedIdentityCredential.GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
   at Azure.Identity.DefaultAzureCredential.GetTokenFromSourcesAsync(TokenCredential[] sources, TokenRequestContext requestContext, Boolean async, CancellationToken cancellationToken)
   at Azure.Identity.DefaultAzureCredential.GetTokenImplAsync(Boolean async, TokenRequestContext requestContext, CancellationToken cancellationToken)
   at Azure.Identity.CredentialDiagnosticScope.FailWrapAndThrow(Exception ex)
   at Azure.Identity.DefaultAzureCredential.GetTokenImplAsync(Boolean async, TokenRequestContext requestContext, CancellationToken cancellationToken)
   at Azure.Identity.DefaultAzureCredential.GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.AccessTokenCache.GetHeaderValueFromCredentialAsync(TokenRequestContext context, Boolean async, CancellationToken cancellationToken)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.AccessTokenCache.GetHeaderValueAsync(HttpMessage message, TokenRequestContext context, Boolean async)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.AccessTokenCache.GetHeaderValueAsync(HttpMessage message, TokenRequestContext context, Boolean async)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.AuthenticateAndAuthorizeRequestAsync(HttpMessage message, TokenRequestContext context)
   at Azure.Security.KeyVault.ChallengeBasedAuthenticationPolicy.AuthorizeRequestInternal(HttpMessage message, Boolean async)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.RedirectPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.HttpPipeline.SendRequestAsync(Request request, CancellationToken cancellationToken)
   at Azure.Security.KeyVault.KeyVaultPipeline.SendRequestAsync(Request request, CancellationToken cancellationToken)
   at Azure.Security.KeyVault.KeyVaultPipeline.GetPageAsync[T](Uri firstPageUri, String nextLink, Func`1 itemFactory, String operationName, CancellationToken cancellationToken)
   at Azure.Core.PageResponseEnumerator.FuncAsyncPageable`1.AsPages(String continuationToken, Nullable`1 pageSizeHint)+MoveNext()
   at Azure.Core.PageResponseEnumerator.FuncAsyncPageable`1.AsPages(String continuationToken, Nullable`1 pageSizeHint)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
   at Azure.AsyncPageable`1.GetAsyncEnumerator(CancellationToken cancellationToken)+MoveNext()
   at Azure.AsyncPageable`1.GetAsyncEnumerator(CancellationToken cancellationToken)+MoveNext()
   at Azure.AsyncPageable`1.GetAsyncEnumerator(CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
   at CCP.PTT.TrustDomainService.Data.Infrastructure.KeyVaultX509CertificateRepository.GetAsync(String name, String thumbprint) in C:\source\CCP.PTT.TrustDomainService.Data.Infrastructure\KeyVaultX509CertificateRepository.cs:line 62
   at CCP.PTT.TrustDomainService.Data.Infrastructure.KeyVaultX509CertificateRepository.GetAsync(String name, String thumbprint) in C:\source\CCP.PTT.TrustDomainService.Data.Infrastructure\KeyVaultX509CertificateRepository.cs:line 62
   at CCP.PTT.TrustDomainService.Controllers.RootCertificatesController.GetAsync(String rootName, String thumbprint) in C:\source\CCP.PTT.TrustDomainService\Controllers\Roots\Certificates\RootCertificatesController.cs:line 117
   at lambda_method250(Closure , Object )
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Logged|12_1(ControllerActionInvoker invoker)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextExceptionFilterAsync>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
```

Here's our code:
```csharp
 public class KeyVaultX509CertificateRepository : IX509CertificateRepository, IDisposable
    {
        private readonly SafeHandle handle = new SafeFileHandle(IntPtr.Zero, true);
        private readonly KeyVaultClient keyVaultClient;
        private readonly string baseUrl;
        private bool disposed;

        /// <summary>
        /// Initializes a new instance of the <see cref="KeyVaultX509CertificateRepository"/> class.
        /// </summary>
        /// <param name="baseUrl">The base url of the keyvault to use.</param>
        public KeyVaultX509CertificateRepository(string baseUrl)
        {
            this.baseUrl = baseUrl;
            CertificateClientOptions certificateClientOptions = new CertificateClientOptions();
            certificateClientOptions.Retry.Mode = RetryMode.Exponential;
            certificateClientOptions.Retry.MaxRetries = 5;
            this.keyVaultClient = new KeyVaultClient(baseUrl, new DefaultAzureCredential(), certificateClientOptions);
        }

        /// <inheritdoc/>
        public async Task<X509Certificate2Collection> GetAsync(string name, string thumbprint)
        {
            if (string.IsNullOrWhiteSpace(name))
            {
                throw new NullOrWhitespaceBadArgumentException(nameof(name));
            }

            if (string.IsNullOrWhiteSpace(thumbprint))
            {
                throw new NullOrWhitespaceBadArgumentException(nameof(thumbprint));
            }

            AsyncPageable<CertificateProperties> certificates = this.keyVaultClient.CertificateClient.GetPropertiesOfCertificateVersionsAsync(name);

            await foreach (CertificateProperties certificate in certificates)
            {
                string certificateThumbprint = BitConverter.ToString(certificate.X509Thumbprint).Replace("-", string.Empty, StringComparison.Ordinal);
                if (string.Equals(certificateThumbprint, thumbprint, StringComparison.OrdinalIgnoreCase))
                {
                    KeyVaultSecret secret = await this.keyVaultClient.SecretClient.GetSecretAsync(name, certificate.Version);

                    byte[] secretBytes = Convert.FromBase64String(secret.Value);
                    X509Certificate2Collection certCollection = new X509Certificate2Collection();

                    if (MeshCertificateService.FormatRootKey(Constants.LinkerdRootName).Equals(name, StringComparison.OrdinalIgnoreCase))
                    {
                        certCollection.Import(secretBytes, string.Empty, X509KeyStorageFlags.MachineKeySet | X509KeyStorageFlags.EphemeralKeySet | X509KeyStorageFlags.Exportable);
                    }
                    else
                    {
                        certCollection.Import(secretBytes, string.Empty, X509KeyStorageFlags.MachineKeySet | X509KeyStorageFlags.PersistKeySet | X509KeyStorageFlags.Exportable);
                    }

                    return certCollection;
                }
            }

            throw new EntityNotFoundException(thumbprint);
        }
}
```

**Environment:**
 - Name and version of the Library package used: Azure.Identity: 1.4.1, Azure.Security.KeyVault.Secrets: 4.2.0, Azure.Security.KeyVault.Certificates: 4.2.0
 - Hosting platform or OS and .NET runtime version (`dotnet --info` output for .NET Core projects):
 .NET SDK (reflecting any global.json):  Version:   5.0.303  Commit:    6409b42649  Runtime Environment:  OS Name:     Windows  OS Version:  10.0.19043  OS Platform: Windows  RID:         win10-x64  Base Path:   C:\Program Files\dotnet\sdk\5.0.303\  Host (useful for support):   Version: 5.0.9   Commit:  208e377a53  .NET SDKs installed:   3.0.103 [C:\Program Files\dotnet\sdk]   3.1.412 [C:\Program Files\dotnet\sdk]   5.0.104 [C:\Program Files\dotnet\sdk]   5.0.302 [C:\Program Files\dotnet\sdk]   5.0.303 [C:\Program Files\dotnet\sdk]  .NET runtimes installed:   Microsoft.AspNetCore.All 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]   Microsoft.AspNetCore.All 2.1.29 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]   Microsoft.AspNetCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]   Microsoft.AspNetCore.App 2.1.29 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]   Microsoft.AspNetCore.App 3.0.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]   Microsoft.AspNetCore.App 3.1.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]   Microsoft.AspNetCore.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]   Microsoft.AspNetCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]   Microsoft.AspNetCore.App 5.0.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]   Microsoft.AspNetCore.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]   Microsoft.NETCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]   Microsoft.NETCore.App 2.1.29 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]   Microsoft.NETCore.App 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]   Microsoft.NETCore.App 3.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]   Microsoft.NETCore.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]   Microsoft.NETCore.App 3.1.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]   Microsoft.NETCore.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]   Microsoft.NETCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]   Microsoft.NETCore.App 5.0.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]   Microsoft.NETCore.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]   Microsoft.WindowsDesktop.App 3.0.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]   Microsoft.WindowsDesktop.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]   Microsoft.WindowsDesktop.App 3.1.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]   Microsoft.WindowsDesktop.App 3.1.18 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]   Microsoft.WindowsDesktop.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]   Microsoft.WindowsDesktop.App 5.0.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]   Microsoft.WindowsDesktop.App 5.0.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
 - IDE and version : [e.g. Visual Studio 16.3]
Visual Studio 16.10.4

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team members best able to assist.

**Comment by christothes** [comment]
Hi @shahriak - Perhaps I'm missing it, but I don't see the details of the exception you received from FailWrapAndThrow. Was there a specific exception message? At a high level from the details provided, it appears that we are failing to get a token from the Managed Identity credential.

One thing you could do to get a bit more detail is to enable logging. 
Details on how to enable logging can be found [here.](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/samples/Diagnostics.md)

A few more questions 
- Does this work locally when `DefaultAzureCredential` falls back to one of the other service principal based credentials such as `AzureCliCredential` or `VisualStudioCredential`?
- Are you able to authenticate to your Key Vault under any scenario using Azure.Identity?

#feedback

**Comment by shahriak** [comment]
@christothes Thank you so much for replying to this thread. The above code was actually unused and therefore, we decided to remove it from our repository. However, I am running into another similar issue, I have added the description here: https://github.com/Azure/azure-sdk-for-net/issues/23381. I appreciate your help!

Does this work locally when DefaultAzureCredential falls back to one of the other service principal based credentials such as AzureCliCredential or VisualStudioCredential?
- Yes, I am able to successfully run all the API and get expected response. Only when I deploy the changes, I see the error message.
Are you able to authenticate to your Key Vault under any scenario using Azure.Identity?
- Yes.
